### PR TITLE
Fix "Non-exhaustive patterns in function nodeOuterHtml"

### LINKED ISSUE #1192: SSC: "Non-exhaustive patterns in function nodeOuterHtml"
I have a component (a simple search box) that crashes whenever I type anything into it. The error goes `IHP/ServerSideComponent/HtmlDiff.hs:(112,1)-(114,46): Non-exhaustive patterns in function nodeOuterHtml`.

The component doesn't crash if I use `show` instead of `intercalate ","`. Here is sample result:

image

```haskell
module Web.Component.Switcher where

import IHP.ServerSideComponent.Types
import IHP.ServerSideComponent.ControllerFunctions
import Web.View.Prelude hiding (query, fetch)
import Web.Controller.Prelude

data Switcher = Switcher {
    switcherQuery :: !Text,
    currentResults :: ![Card]
}

data SwitcherController = SetSwitcherQuery {switcherQuery :: Text}
    deriving (Eq, Show, Data)

$(deriveSSC ''SwitcherController)

instance Component Switcher SwitcherController where
    initialState = Switcher {switcherQuery = "", currentResults = []}

    render Switcher{switcherQuery, currentResults} = [hsx|
        <div>
            <input type="text" 
                value={switcherQuery}
                onkeyup="callServerAction('SetSwitcherQuery', { switcherQuery: this.value })"
            >
            <span>{intercalate ", " $ map (get #title) currentResults}</span>
        </div>
    |]

    action state SetSwitcherQuery {switcherQuery} = do
        results <- query @Card
            |> filterWhereILike (#title, "%" <> switcherQuery <> "%")
            |> fetch
        pure $ state {
            switcherQuery = switcherQuery, 
            currentResults = results
            }
```

### PR DESCRIPTION
**Pull request by neongreen**
Fixes #1192.

Before this is merged, I'd love to know if there is also a testsuite for the client-side HTML patching code.

### CODE REVIEW

File: IHP/ServerSideComponent/HtmlDiff.hs [hunk 1]
```diff
@@ -73,7 +73,7 @@ diffNodes' path Children { children = oldChildren } Children { children = newChi
             patchElements (new:nextNewNode:newRest) (old:oldRest) !index | (not $ new `isNodeEqIgnoringPosition` old) && (old `isNodeEqIgnoringPosition` nextNewNode) = [ CreateNode { html = nodeOuterHtml new ?newHtml, path = (index:path) } ] <> (patchElements (newRest) (oldRest) (index + 2)) -- [A, C <old>] -> [A, B <new>, C <nextNewNode>]
             patchElements (new:newRest) (old:nextOld:oldRest) !index | (not $ new `isNodeEqIgnoringPosition` old) && (new `isNodeEqIgnoringPosition` nextOld) = [ DeleteNode { node = old, path = (index:path) } ] <> (patchElements (newRest) (oldRest) (index + 1)) -- [A, B <old>, C <nextOldNode> ] -> [A, C <new>]
             patchElements (new:newRest) (old:oldRest) !index = (diffNodes' (index:path) old new) <> (patchElements newRest oldRest (index + 1))
-            patchElements (new:newRest) oldRest !index = [ CreateNode { html = nodeOuterHtml new ?newHtml, path = (index:path) } ] <> (patchElements newRest oldRest (index + 1))
+            patchElements (new:newRest) [] !index = [ CreateNode { html = nodeOuterHtml new ?newHtml, path = (index:path) } ] <> (patchElements newRest [] (index + 1))
```
Comment (neongreen) [inline IHP/ServerSideComponent/HtmlDiff.hs:76]: Replaced with `[]` because it's always true and seems more informative.

**Comment by mpscholten** [comment]
There's no test suite for the JS sadly

**Comment by neongreen** [comment]
I tried it locally and it seems to work.

**Comment by mpscholten** [comment]
Thanks! I could reproduce the issue and after apply the patch it was fixed 👍 Also didn't see anything that was broken after applying the patch

**Review by mpscholten — APPROVED** [review]

**Pull request merged**